File: env/drone.c
// Standalone C demo for drone environment
// Compile using: ./scripts/build_ocean.sh drone [local|fast]
// Run with: ./drone

#include "drone.h"
#include "puffernet.h"
#include "render.h"
#include <time.h>

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#endif

double randn(double mean, double std) {
    static int has_spare = 0;
    static double spare;

    if (has_spare) {
        has_spare = 0;
        return mean + std * spare;
    }

    has_spare = 1;
    double u, v, s;
    do {
        u = 2.0 * rand() / RAND_MAX - 1.0;
        v = 2.0 * rand() / RAND_MAX - 1.0;
        s = u * u + v * v;
    } while (s >= 1.0 || s == 0.0);

    s = sqrt(-2.0 * log(s) / s);
    spare = v * s;
    return mean + std * (u * s);
}

#ifndef LINEAR_DIM
#define LINEAR_DIM 64
#endif

#ifndef LSTM_DIM
#define LSTM_DIM 16
#endif

typedef struct LinearContLSTM LinearContLSTM;
struct LinearContLSTM {
    int num_agents;
    float* obs;
    int num_actions;
    float* log_std;
    Linear* encoder1;
    GELU*  gelu1;
    Linear* encoder2;
    GELU*  gelu2;
    LSTM* lstm;
    Linear* actor;
    Linear* value_fn;
};

LinearContLSTM* make_linearcontlstm(Weights* weights, int num_agents, int input_dim,
                                    int logit_sizes[], int num_actions) {
    LinearContLSTM* net = calloc(1, sizeof(LinearContLSTM));
    net->num_agents = num_agents;
    net->obs = calloc(num_agents * input_dim, sizeof(float));
    net->num_actions = logit_sizes[0];
    
    // Must match export order exactly:
    net->log_std  = get_weights(weights, net->num_actions);                      // 1. decoder_logstd
    net->encoder1 = make_linear(weights, num_agents, input_dim, LINEAR_DIM);     // 2-3. encoder.0
    net->gelu1    = make_gelu(num_agents, LINEAR_DIM);
    net->encoder2 = make_linear(weights, num_agents, LINEAR_DIM, LSTM_DIM);      // 4-5. encoder.2
    net->gelu2    = make_gelu(num_agents, LSTM_DIM);
    net->actor    = make_linear(weights, num_agents, LSTM_DIM, net->num_actions);// 6-7. decoder_mean
    net->value_fn = make_linear(weights, num_agents, LSTM_DIM, 1);               // 8-9. value (FIX: was LSTM_DIM+4)
    net->lstm     = make_lstm(weights, num_agents, LSTM_DIM, LSTM_DIM);          // 10-13. lstm

    return net;
}

void free_linearcontlstm(LinearContLSTM* net) {
    free(net->obs);
    free(net->encoder1);
    free(net->gelu1);
    free(net->encoder2);
    free(net->gelu2);
    free(net->lstm);
    free(net->actor);
    free(net->value_fn);
    free(net);
}

void forward_linearcontlstm(LinearContLSTM* net, float* observations) {
    linear(net->encoder1, observations);
    gelu(net->gelu1, net->encoder1->output);
    linear(net->encoder2, net->gelu1->output);
    gelu(net->gelu2, net->encoder2->output);
    lstm(net->lstm, net->gelu2->output);
    linear(net->actor, net->lstm->state_h);
}

void sample_linearcontlstm(LinearContLSTM* net, float* actions, int deterministic) {
    for (int b = 0; b < net->num_agents; b++) {
        for (int i = 0; i < net->num_actions; i++) {
            int idx = b * net->num_actions + i;
            float mean = net->actor->output[idx];
            if (deterministic) {
                actions[idx] = mean;
            } else {
                float std = expf(net->log_std[i]);
                actions[idx] = (float)randn(mean, std);
            }
        }
    }
}

void generate_dummy_actions(DroneEnv* env) {
    // Generate random floats in [-1, 1] range
    env->actions[0] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
    env->actions[1] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
    env->actions[2] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
    env->actions[3] = ((float)rand() / (float)RAND_MAX) * 2.0f - 1.0f;
}

#ifdef __EMSCRIPTEN__
typedef struct {
    DroneEnv* env;
    LinearContLSTM* net;
    Weights* weights;
} WebRenderArgs;

void emscriptenStep(void* e) {
    WebRenderArgs* args = (WebRenderArgs*)e;
    DroneEnv* env = args->env;
    LinearContLSTM* net = args->net;

    for (int i = 0; i < env->num_agents; i++) {
        int base = i * obs_size;
        env->observations[base + 19] = 0.0f;
        env->observations[base + 20] = 0.0f;
        env->observations[base + 21] = 0.0f;
        env->observations[base + 22] = 0.0f;
    }

    forward_linearcontlstm(net, env->observations);
    sample_linearcontlstm(net, env->actions, 0);
    c_step(env);
    c_render(env);
    return;
}

WebRenderArgs* web_args = NULL;
#endif

int main() {
    srand(time(NULL)); // Seed random number generator

    DroneEnv* env = calloc(1, sizeof(DroneEnv));
    size_t obs_size = 23;
    size_t act_size = 4;

    env->num_agents = 64;
    env->max_rings = 10;
    env->task = HOVER;
    env->hover_target_dist = 0.5f;
    env->hover_dist = 0.05f;
    env->hover_omega = 0.05;
    env->hover_vel = 0.01;
    env->num_obs = obs_size;
    env->num_envs = 1;
    env->env_index = 0;
    init(env);

    env->observations = (float*)calloc(env->num_agents * obs_size, sizeof(float));
    env->actions = (float*)calloc(env->num_agents * act_size, sizeof(float));
    env->rewards = (float*)calloc(env->num_agents, sizeof(float));
    env->terminals = (unsigned char*)calloc(env->num_agents, sizeof(float));

    Weights* weights = load_weights("resources/drone/puffer_drone_weights.bin", 4841);
    int logit_sizes[1] = {4};
    LinearContLSTM* net = make_linearcontlstm(weights, env->num_agents, obs_size, logit_sizes, 1);

    if (!env->observations || !env->actions || !env->rewards) {
        fprintf(stderr, "ERROR: Failed to allocate memory for demo buffers.\n");
        free(env->observations);
        free(env->actions);
        free(env->rewards);
        free(env->terminals);
        free(env);
        return 0;
    }

    init(env);
    c_reset(env);

#ifdef __EMSCRIPTEN__
    WebRenderArgs* args = calloc(1, sizeof(WebRenderArgs));
    args->env = env;
    args->net = net;
    args->weights = weights;
    web_args = args;

    emscripten_set_main_loop_arg(emscriptenStep, args, 0, true);
#else
    c_render(env);

    while (!WindowShouldClose()) {
        for (int i = 0; i < env->num_agents; i++) {
            int base = i * obs_size;
            env->observations[base + 19] = 0.0f;
            env->observations[base + 20] = 0.0f;
            env->observations[base + 21] = 0.0f;
            env->observations[base + 22] = 0.0f;
        }
        forward_linearcontlstm(net, env->observations);
        sample_linearcontlstm(net, env->actions, 0);
        c_step(env);
        c_render(env);
    }

    c_close(env);
    free_linearcontlstm(net);
    free(env->observations);
    free(env->actions);
    free(env->rewards);
    free(env->terminals);
    free(env);
#endif

    return 0;
}

File: env/drone.py
import numpy as np
import gymnasium

import pufferlib
from pufferlib.ocean.drone import binding


# python bindings for drone env
class Drone(pufferlib.PufferEnv):
    def __init__(
        self,
        task="HOVER",
        num_envs=16,
        num_drones=64,
        max_rings=5,
        alpha_hover=1.0,
        alpha_shaping=0.001,
        alpha_omega=0.0001,
        alpha_dist=1.0,
        hover_target_dist=10.0,
        hover_dist=0.1,
        hover_omega=0.1,
        hover_vel=0.1,
        render_mode=None,
        report_interval=1024,
        buf=None,
        seed=0,
    ):
        num_obs = 23
        self.single_observation_space = gymnasium.spaces.Box(
            low=-1,
            high=1,
            shape=(num_obs,),
            dtype=np.float32,
        )

        self.single_action_space = gymnasium.spaces.Box(
            low=-1, high=1, shape=(4,), dtype=np.float32
        )

        self.num_agents = num_envs * num_drones
        self.render_mode = render_mode
        self.report_interval = report_interval
        self.tick = 0

        super().__init__(buf)
        self.actions = self.actions.astype(np.float32)

        c_envs = []
        for i in range(num_envs):
            c_envs.append(
                binding.env_init(
                    self.observations[i * num_drones : (i + 1) * num_drones],
                    self.actions[i * num_drones : (i + 1) * num_drones],
                    self.rewards[i * num_drones : (i + 1) * num_drones],
                    self.terminals[i * num_drones : (i + 1) * num_drones],
                    self.truncations[i * num_drones : (i + 1) * num_drones],
                    i,
                    task=task,
                    num_agents=num_drones,
                    max_rings=max_rings,
                    env_index=i,
                    num_envs=num_envs,
                    alpha_hover=alpha_hover,
                    alpha_shaping=alpha_shaping,
                    alpha_omega=alpha_omega,
                    alpha_dist=alpha_dist,
                    hover_target_dist=hover_target_dist,
                    hover_dist=hover_dist,
                    hover_omega=hover_omega,
                    hover_vel=hover_vel,
                    num_obs=num_obs,
                )
            )

        self.c_envs = binding.vectorize(*c_envs)

    def reset(self, seed=None):
        self.tick = 0
        binding.vec_reset(self.c_envs, seed)
        return self.observations, []

    def step(self, actions):
        self.actions[:] = actions

        self.tick += 1
        binding.vec_step(self.c_envs)

        info = []
        if self.tick % self.report_interval == 0:
            log_data = binding.vec_log(self.c_envs)
            if log_data:
                info.append(log_data)

        return (self.observations, self.rewards, self.terminals, self.truncations, info)

    def render(self):
        binding.vec_render(self.c_envs, 0)

    def close(self):
        binding.vec_close(self.c_envs)


def test_performance(timeout=10, atn_cache=1024):
    env = Drone(num_envs=1000)
    env.reset()
    tick = 0

    actions = [env.action_space.sample() for _ in range(atn_cache)]

    import time

    start = time.time()
    while time.time() - start < timeout:
        atn = actions[tick % atn_cache]
        env.step(atn)
        tick += 1

    print(f"SPS: {env.num_agents * tick / (time.time() - start)}")


if __name__ == "__main__":
    test_performance()
